File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_11/CheckpointTest.java
package com.fResult.reactor.ch5_11;

import lombok.SneakyThrows;
import lombok.extern.log4j.Log4j2;
import lombok.val;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Flux;
import reactor.test.StepVerifier;

import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.time.Duration;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.assertTrue;

@Log4j2
class CheckpointTest {
  @Test
  void checkpoint() {
    val stackTrace = new AtomicReference<String>();
    val checkpoint =
        Flux.error(new IllegalArgumentException("Oops"))
            .checkpoint()
            .delayElements(Duration.ofMillis(1));

    StepVerifier.create(checkpoint)
        .expectErrorMatches(
            ex -> {
              stackTrace.set(stackTraceToString(ex));
              return ex instanceof IllegalArgumentException;
            })
        .verify();

    assertTrue(stackTrace.get().contains("Error has been observed at the following site(s):"));
  }

  @SneakyThrows(IOException.class)
  private String stackTraceToString(Throwable ex) {
    try (val sw = new StringWriter();
        val pw = new PrintWriter(sw)) {

      ex.printStackTrace(pw);
      return sw.toString();
    }
  }
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_11/CheckpointTest.kt
package com.fResult.reactor.ch5_11

import reactor.core.publisher.Flux
import reactor.test.StepVerifier
import java.io.PrintWriter
import java.io.StringWriter
import java.util.concurrent.atomic.AtomicReference
import kotlin.test.Test
import kotlin.test.assertTrue
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.toJavaDuration

class CheckpointTest {
  @Test
  fun checkpoint() {
    val stackTrace = AtomicReference<String>()

    val checkpoint = Flux.error<String>(IllegalArgumentException("Oops!"))
      .checkpoint()
      .delayElements(1.milliseconds.toJavaDuration())

    StepVerifier.create(checkpoint)
      .expectErrorMatches { ex ->
        stackTrace.set(stackTraceToString(ex))

        ex is IllegalArgumentException
      }
      .verify()

    assertTrue(stackTrace.get().contains("Error has been observed at the following site(s):"))
  }

  private fun stackTraceToString(ex: Throwable) = StringWriter().use(stackTraceWriteHandler(ex))

  private fun stackTraceWriteHandler(ex: Throwable): (StringWriter) -> String = { sw ->
    PrintWriter(sw).use(captureStackTraceToString(ex, sw))
  }

  private fun captureStackTraceToString(ex: Throwable, sw: StringWriter): (PrintWriter) -> String = { pw ->
    ex.printStackTrace(pw)
    sw.toString()
  }
}
